pkg/solana/chainwriter/transform_registry.go

```diff
@@ -80,35 +90,164 @@ func CCIPExecuteArgsTransform(ctx context.Context, client client.MultiClient, ar
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to calculate compute unit limit: %w", err)
+	}
+
+	options = append(options, txmutils.SetEstimateComputeUnitLimit(false), txmutils.SetComputeUnitLimit(computeUnits))
+
+	if len(argsTransformed.Info.AbstractReports) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of reports, got %d, expect 1", len(argsTransformed.Info.AbstractReports))
+	}
+	report := argsTransformed.Info.AbstractReports[0]
+	if len(report.Messages) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of messages, got %d, expect 1", len(report.Messages))
+	}
+
+	message := report.Messages[0]
+	sourceChainSel := message.Header.SourceChainSelector
+	if len(argsTransformed.Info.MerkleRoots) != 1 {
+		return nil, nil, nil, nil, fmt.Errorf("encountered unexpected number of merkle roots, got %d, expect 1", len(argsTransformed.Info.MerkleRoots))
+	}
+	merkleRoot := argsTransformed.Info.MerkleRoots[0].MerkleRoot
+
+	var messageAccounts []ccip_offramp.CcipAccountMeta
+	if !message.Receiver.IsZeroOrEmpty() {
+		logicReceiver := solana.PublicKeyFromBytes(message.Receiver)
+		// Append logic receiver as the first messaging account for derivation
+		messageAccounts = append(messageAccounts, ccip_offramp.CcipAccountMeta{
+			Pubkey:     logicReceiver,
+			IsSigner:   false,
+			IsWritable: false,
+		})
+		// Extract the user defined accounts
+		userAccountsLookup := AccountLookup{
+			Name:       "UserAccounts",
+			Location:   "ExtraData.ExtraArgsDecoded.accounts",
+			IsWritable: MetaBool{BitmapLocation: "ExtraData.ExtraArgsDecoded.accountIsWritableBitmap"},
+			IsSigner:   MetaBool{Value: false},
+		}
+		userAccounts, resolveErr := userAccountsLookup.Resolve(args)
+		// If err is ErrLookupNotFoundAtLocation, allow process to continue in case accounts are not needed
+		if resolveErr != nil && !errors.Is(resolveErr, ErrLookupNotFoundAtLocation) {
+			return nil, nil, nil, nil, fmt.Errorf("failed to resolve user accounts: %w", resolveErr)
+		}
+		messageAccounts = append(messageAccounts, ConvertToCCIPAccountMetas(userAccounts)...)
+	}
+
+	// Extract token transfers
+	var tokenTransfers []ccip_offramp.TokenTransferAndOffchainData
+	var tokenReceiver solana.PublicKey
+	var messageTokenData [][]byte
+	if len(message.TokenAmounts) > 0 {
+		tokenTransfers = make([]ccip_offramp.TokenTransferAndOffchainData, 0, len(message.TokenAmounts))
+		if len(argsTransformed.ExtraData.DestExecDataDecoded) != len(message.TokenAmounts) {
+			return nil, nil, nil, nil, fmt.Errorf("unexpected number of DestExecData encountered. expect the same number as token transfers %d, got %d", len(message.TokenAmounts), len(argsTransformed.ExtraData.DestExecDataDecoded))
+		}
+		// If message contains token transfers, extract offchain token data for the message if any exists
+		if len(report.OffchainTokenData) > 0 {
+			messageTokenData = report.OffchainTokenData[0]
+		}
```

Comment:
I know we expect the OffchainTokenData to have a single entry, as there is up to 1 token transfer per message today (and 1 message per report), but it feels like something that should be validated here and throw an unexpected number of... error if the assumption does not hold

Reply:
Ya good point! If I'm making this assumption, I should make some noise if it doesn't hold .